drs_downloader/manager.py

```diff
@@ -0,0 +1,393 @@
+                logger.error(f"{drs_object.name} had missing part.")
+                return drs_object
+
+            paths.extend(chunk_paths)
+
+        drs_object.file_parts = paths
+
+        i = 1
+        filename = f"{drs_object.name}"
+        original_file_name = Path(filename)
+        while True:
+            if os.path.isfile(destination_path.joinpath(filename)):
+                filename = f"{original_file_name}({i})"
+                i = i + 1
+                continue
+            break
+
+        # re-assemble and test the file parts
+        # hash function dynamic
+        checksum_type = drs_object.checksums[0].type
+        assert checksum_type in hashlib.algorithms_available, f"Checksum {checksum_type} not supported."
+        checksum = hashlib.new(checksum_type)
+        with open(destination_path.joinpath(filename), 'wb') as wfd:
+            # sort the items of the list in place - Numerically based on start i.e. "xxxxxx.start.end.part"
+            drs_object.file_parts.sort(key=lambda x: int(str(x).split('.')[-3]))
+            for f in drs_object.file_parts:
+                fd = open(f, 'rb')
+                wrapped_fd = Wrapped(fd, checksum)
+                # efficient way to write
+                shutil.copyfileobj(wrapped_fd, wfd)
+                # explicitly close all
+                wrapped_fd.close()
+                fd.close()
+                wfd.flush()
+        actual_checksum = checksum.hexdigest()
+
+        actual_size = os.stat(Path(destination_path.joinpath(filename))).st_size
+
+        # compare calculated md5 vs expected
+        expected_checksum = drs_object.checksums[0].checksum
+        if expected_checksum != actual_checksum:
+            msg = f"Actual {checksum_type} hash {actual_checksum} does not match expected {expected_checksum}"
+            drs_object.errors.append(msg)
+
+        if drs_object.size != actual_size:
+            msg = f"The actual size {actual_size} does not match expected size {drs_object.size}"
+            drs_object.errors.append(msg)
+
+        # if error leave parts in place for now
+        if len(drs_object.errors) == 0:
+            # ok, so clean up file parts
+            for f in drs_object.file_parts:
+                f.unlink()
+
+        return drs_object
+
+    async def _run_download(self, drs_objects: List[DrsObject], destination_path: Path) -> List[DrsObject]:
+        """
+        Create tasks to sign and download, display progress.
+        Args:
+            drs_objects: list of drs objects to download
+
+        Returns:
+            updated list of drs objects
+        """
+
+        # first sign the urls
+        tasks = []
+        for drs_object in drs_objects:
+            task = asyncio.create_task(self._drs_client.sign_url(drs_object=drs_object))
+            tasks.append(task)
+
+        drs_objects_with_signed_urls = [
+            await f for f in asyncio.as_completed(tasks)
+        ]
+
+        # second, download the parts
```

Comment:
I think for really large files (many tens or hundreds of GB), the signed URL will expire before all of the parts have at least started downloading. We can test this to verify my understanding based on reading the code, yet I am pretty sure that is true.

The TDR DRS signed URLs expire in 15 minutes, and Gen3 DRS signed URLs expire in 60 minutes, yet I am pretty sure that for file sizes in the range of hundreds of GB the signed URL expiration would occur even with the 60-minute timeout.

Reply:
@mbaumann-broad Do we know what the http status will be in this scenario?  Address in separate PR #19